### DISCUSSION TITLE
Is a typst file safe?

### DISCUSSION THREAD

**Discussion by YDX-2147483647** [typst/typst#2470] [Q&A] [answered]
Can a typst file run arbitrary code? Is it safe to `typst compile` an untrusted `*.typ`?

Typst can [load data from external files](https://typst.app/docs/reference/data-loading/), but will it lead to a virus?

(I don't mean `unsafe { }` in rust)

---

I make some research and find a [paper](https://hovav.net/ucsd/papers/csr10.html) about LaTeX:

> We show that malicious TeX, BibTeX, and METAPOST files can lead to arbitrary code execution, viral infection, denial of service, and data exfiltration, through the file I/O capabilities exposed by TeX’s Turing-complete macro language.

**Comment by Enter-tainer** [comment]
afaik running typst code won't led to arbitrary code execution, or it is a security bug.

**Reply by Dherse** [reply]
but but, with `shell-escape` we could make typst unsafe! That would be 100x better /s

**Reply by YDX-2147483647** [reply]
Do you mean https://github.com/elteammate/typst-shell-escape?

**Reply by Enter-tainer** [reply]
> It mounds a custom userspace filesystem. The only way Typst can interact with the outer world is by reading from files, and we are using this to our advantage.

the repo you posted explains how this works. this may give you some clue of how "safe" typst is

**Reply by YDX-2147483647** [reply]
I see… Thank you!

**Comment by laurmaedje** [comment, marked as answer]
As @Enter-tainer said, we take the safety of just compiling a Typst file very seriously. Features that would compromise safety (like writing files, shell escape, even reading files outside of the working/project directory, etc.) won't be accepted into the Typst compiler.
